File: 104 Maximum Depth of Binary Tree/Maximum Depth of Binary Tree.cpp
// Recursive DFS (Depth-First Search)
int maxDepth(TreeNode *root)
{
    if (root == nullptr)
    {
        return 0;
    }
    return max(maxDepth(root->left), maxDepth(root->right)) + 1;
}

// BFS (Breadth-First Search)
int maxDepth(TreeNode *root)
{
    if (root == nullptr)
    {
        return 0;
    }
    queue<TreeNode *> q;
    q.push(root);
    int level = 0;

    while (!q.empty())
    {
        int count = q.size();
        for (int i = 0; i < count; i++)
        {
            TreeNode *node = q.front();
            q.pop();

            if (node->left != nullptr)
            {
                q.push(node->left);
            }
            if (node->right != nullptr)
            {
                q.push(node->right);
            }
        }
        level++;
    }
    return level;
}

File: 104 Maximum Depth of Binary Tree/Maximum Depth of Binary Tree.py
# Python 3'ü seçmeyi unutma =)
# Recursive DFS (Depth-First Search)
class Solution:
    def maxDepth(self, root: TreeNode) -> int:
        if not root:
            return 0
        return max(self.maxDepth(root.left), self.maxDepth(root.right)) + 1


# BFS (Breadth-First Search)
class Solution:
    def maxDepth(self, root: Optional[TreeNode]) -> int:
        # BFS
        if not root:
            return 0
        queue = [root]
        cnt = 0
        while queue:
            temp_queue = queue
            queue = []
            for node in temp_queue:
                if node.left:
                    queue.append(node.left)
                if node.right:
                    queue.append(node.right)
            cnt += 1
        return cnt
